core/src/main/java/org/apache/calcite/rex/RexBuilder.java

```diff
@@ -1918,31 +1918,24 @@ public RexLiteral makeZeroLiteral(RelDataType type) {
   }
 
   public RexNode makeZeroForNestedType(RelDataType type) {
+    SqlOperator op;
     switch (type.getSqlTypeName()) {
     case ARRAY:
-      ArraySqlType arrayType = (ArraySqlType) type;
-      RexNode zeroElement = makeZeroForNestedType(arrayType.getComponentType());
-      return makeCall(type, SqlStdOperatorTable.ARRAY_VALUE_CONSTRUCTOR,
-          ImmutableList.of(zeroElement));
+      op = SqlStdOperatorTable.ARRAY_VALUE_CONSTRUCTOR;
```

Comment:
What is the type of the constructed array?
I wonder whether this will create problems because the type needs to be inferred.
But you know the type already, so you can synthesize a `CAST(ARRAY[] AS type)`.
Same below.

Reply:
When using CAST, you may need to write something like ARRAY[INTEGER]

Reply:
I am not sure what this statements means. This is not legal SQL.
And here I am talking about the internal representation.

Reply:
You mean to keep the actual type and the writing consistent to avoid problems with `CAST(ARRAY[] AS type)` in SQL, right? I'll think about it and may need to adjust it again.

Reply:
This should be a simple change. Just wrap the constructor call into a cast that produces `type`.

Reply:
Thanks! I will fix it soon.